nixio/cmd/nixexplore.py

```diff
@@ -0,0 +1,622 @@
+                                           description="Search for metadata items or display metadata (sub)trees.")
+    meta_parser.add_argument("-p", "--pattern", type=str, default=[], nargs="+", help=mdata_pattern_help)
+    meta_parser.add_argument("-d", "--depth", type=int, default=-1,
+                             help="maximum depth of metadata tree output, default is %(default)s, full depth")
+    add_default_args(meta_parser)
+    add_default_file_args(meta_parser)
+    meta_parser.set_defaults(func=mdata_worker)
+    # add value search?
+    # add option to specify directly if one looks for a property which would increase performance
+
+
+def create_data_parser(parent_parser):
+    data_parser = parent_parser.add_parser("data", help="Search and display information about data entities",
+                                           description=data_parser_help)
+    data_parser.add_argument("-p", "--pattern", type=str, help=data_pattern_help)
+    add_default_args(data_parser)
+    add_default_file_args(data_parser)
+    data_parser.set_defaults(func=data_worker)
+
+
+def create_dump_parser(parent_parser):
+    dump_parser = parent_parser.add_parser("dump", help="Dump stored data to stdout",
+                                           description=dump_parser_help)
+    dump_parser.add_argument("-p", "--pattern", type=str, help=data_pattern_help)
+    add_default_args(dump_parser)
+    add_default_file_args(dump_parser)
+    dump_parser.set_defaults(func=dump_worker)
+
+
+def create_plot_parser(parent_parser):
+    if not nw_present:
+        return
+    plot_parser = parent_parser.add_parser("plot", help="Create basic plots of stored data.",
+                                           description=plot_parser_help)
+    plot_parser.add_argument("-p", "--pattern", type=str, help=data_pattern_help)
+    add_default_args(plot_parser)
+    add_default_file_args(plot_parser)
+    plot_parser.set_defaults(func=plot_worker)
+
+
+def create_file_parser(parent_parser):
+    file_parser = parent_parser.add_parser("file", help="Display basic file info",
+                                           description="Quick display of file information such as " +
+                                           "creation date, file size and structure etc.")
+    file_parser.add_argument("-v", "--verbosity", action="count",
+                             help="increase output verbosity, use -v, -vv, -vvv for more verbose output")
+    add_default_file_args(file_parser)
+    file_parser.set_defaults(func=file_worker)
+
+
+def create_parser():
+    parser = argparse.ArgumentParser(prog="nixio-explore",
+                                     description=general_help)
+    subparsers = parser.add_subparsers(title="commands",
+                                       help="Sub commands for working on data and metadata",
+                                       description=tool_description)
```

Comment:
There's a small issue here with running `nixio-explore` without a subcommand.  Since it's not marked as required, the `func` attribute isn't set if no subcommand is specified and the `args.func(args)` in main throws an exception.

Adding `required=True` is one solution, but that was added in Python 3.7, so if we want to support 3.6 it's not an option.  The other solution is to add `dest="func"` to `add_subparsers()`, which will create the `func` attribute every time and then in main we can do something like
```python
    parser = create_parser()
    args = parser.parse_args()
    if not args.func:
        parser.print_help()
        sys.exit(1)
    args.func(args)
```

Reply:
thanks, done